### DISCUSSION TITLE
Is it possible to implement our a PSO optimiser?

### DISCUSSION THREAD

**Discussion by peterstav1** [mitsuba-renderer/mitsuba3#272] [Q&A]
Is it possible to implement our own custom PSO optimizer and if so how do we begin with something like that?

**Comment by Speierers** [comment]
Although this should be feasible, it looks like you will need to render the frames for different scene state (e.g. particles in the PSO) before taking a gradient step. I don't really see a way to evaluate the function (e.g. rendering) for all those particles at once.

**Reply by peterstav1** [reply]
Could you not have multiple copies (dictionaries) of the same scene (one for each particle) and then evaluate the function (rendering) for each particle separately , before selecting which particle 'to follow' after a number of optimization steps?

**Reply by Speierers** [reply]
I believe it would be very difficult to achieve this in Mitsuba 3 unfortunately.
